pkg/generic/thrift/json_go116plus_amd64.go

```diff
@@ -0,0 +1,152 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package thrift
+
+import (
+	"context"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/bytedance/gopkg/lang/mcache"
+	"github.com/cloudwego/dynamicgo/conv/j2t"
+
+	"github.com/cloudwego/kitex/pkg/generic/descriptor"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	cthrift "github.com/cloudwego/kitex/pkg/remote/codec/thrift"
+	"github.com/cloudwego/kitex/pkg/utils"
+)
+
+// Write write json string to out thrift.TProtocol
+func (m *WriteJSON) Write(ctx context.Context, out thrift.TProtocol, msg interface{}, requestBase *Base) error {
+	// fallback logic
+	if !m.dynamicgoEnabled {
+		return m.originalWrite(ctx, out, msg, requestBase)
+	}
+
+	var cv j2t.BinaryConv
+	// dynamicgo logic
+	if !m.hasRequestBase {
+		requestBase = nil
+	}
+	if requestBase != nil {
+		cv = m.j2tBinaryConvWithThriftBase
+	} else {
+		cv = m.j2tBinaryConv
+	}
+
+	// msg is void
+	if _, ok := msg.(descriptor.Void); ok {
+		if err := m.writeHead(out); err != nil {
+			return err
+		}
+		if err := m.writeFields(ctx, out, Void, nil, nil, nil); err != nil {
+			return err
+		}
+		return writeTail(out)
+	}
+
+	// msg is string
+	s, ok := msg.(string)
+	if !ok {
+		return perrors.NewProtocolErrorWithType(perrors.InvalidData, "decode msg failed, is not string")
+	}
+
+	transBuff := utils.StringToSliceByte(s)
+	dbuf := mcache.Malloc(len(transBuff))[0:0]
+	defer mcache.Free(dbuf)
+
+	if err := m.writeHead(out); err != nil {
+		return err
+	}
+	if err := m.writeFields(ctx, out, String, &cv, transBuff, dbuf); err != nil {
+		return err
+	}
+	return writeTail(out)
+}
+
+type MsgType int
+
+const (
+	Void MsgType = iota
+	String
+)
+
+func (m *WriteJSON) writeFields(ctx context.Context, out thrift.TProtocol, msgType MsgType, cv *j2t.BinaryConv, transBuff, dbuf []byte) error {
+	for _, field := range m.dynamicgoTypeDsc.Struct().Fields() {
+		// Exception field
+		if !m.isClient && field.ID() != 0 {
+			// generic server ignore the exception, because no description for exception
+			// generic handler just return error
+			continue
+		}
+
+		if err := out.WriteFieldBegin(field.Name(), field.Type().Type().ToThriftTType(), int16(field.ID())); err != nil {
+			return err
+		}
+		switch msgType {
+		case Void:
+			if err := out.WriteStructBegin(field.Name()); err != nil {
+				return err
+			}
+			if err := out.WriteFieldStop(); err != nil {
+				return err
+			}
+			if err := out.WriteStructEnd(); err != nil {
+				return err
+			}
+		case String:
+			// encode using dynamicgo
+			// json []byte to thrift []byte
+			if err := cv.DoInto(ctx, field.Type(), transBuff, &dbuf); err != nil {
+				return err
+			}
+		}
```

Comment:
all possible msgType are Void and String here?

Reply:
@felix021 yes. If there's another type, I throw error on L67